### ISSUE TITLE
[Forms] Extracting Defaults form a schema

### ISSUE THREAD

**Issue by logaretm** [standard-schema/standard-schema#11]
Hello, I'm the creator and maintainer of https://github.com/logaretm/vee-validate and I'm very much interested in this project.

I have a few ideas from the perspective of a long-time maintainer of a popular form library that makes use of schema validation for Zod, Valibot, Yup and others.

I will keep each idea to one issue at a time to not muddy the discussion.

I have come up a [Typed Schema concept](https://vee-validate.logaretm.com/v4/guide/composition-api/typed-schema) with similar aims to what you have here to address some of the issues you mentioned in the README.

## What

One thing comes to mind is defaults and how can form libraries take advantage of that information to initialize fields with default/initial values.

Take the following snippets as an example, we can define a few fields with Zod and Valibot with default values like this:

#### Zod

```ts
z.object({
  name: z.string().default('test'),
  age: z.number().default(11),
});
```

#### Valibot

```ts
v.object({
  age: v.optional(v.number(), 11),
});
```

AFAIK Valibot offers utilities to extract those values [like we do here](https://github.com/logaretm/vee-validate/blob/main/packages/valibot/src/index.ts#L57-L71), we also do that with Zod with a [hacky approach](https://github.com/logaretm/vee-validate/blob/main/packages/zod/src/index.ts#L45-L54).

In vee-validate at least there are a few hard requirements for being able to resolve initial values, but the most important one is schemas must be `sync`. Since forms need to have their values initialized on mount rather than at unknown point in the future.

## How (Suggestion)

I believe extending the interface with an optional `~defaults` might prove very useful for form libraries.

```ts
export interface StandardSchema<Input = unknown, Output = Input> {
  // ...

  /**
   * Resolves the defined initial values for the schema.
   */
  readonly "~defaults": StandardDefaults<TInput>;
}
```

What is hard to answer here is:

> what is the type of `~defaults` return type? is it TInput or TOutput?

I honestly don't know, in reality it is a mix of both in real-world forms. But usually I would say it fits `TInput` more.

What do you guys think? Should Standard schema accommodate such cases or should there be an extended version of the standard that caters more to form libraries?

**Comment by ssalbdivad** [comment]
In ArkType, defaults must conform to the input type so that any transformations can be applied to them as if the value were provided directly, so I'd be in favor of that.

What I think is the tougher question here is how `~defaults` would be organized. One option would be to just mirror the shape of the props on the schema, but it duplicates those names and critically, is incompatible with unions.

Realistically, we'd need to address https://github.com/standard-schema/standard-schema/issues/7 and have an established representation of a schema's AST before we can make good decisions about how to associate default values with it.

**Comment by fabian-hiller** [comment]
> In ArkType, defaults must conform to the input type so that any transformations can be applied to them as if the value were provided directly, so I'd be in favor of that.

I agree. In Valibot we have the concept of defaults and fallbacks. Defaults are used when no input is given and they always have to match the input type for the same reason as in ArkType. Fallbacks always match the output as they are returned directly in case of validation errors.

> Realistically, we'd need to address #7 and have an established representation of a schema's AST before we can make good decisions about how to associate default values with it.

I think I agree. We would include defaults elsewhere in this AST similar to how it works in JSON Schema. Valibot already includes a custom AST, probably similar to how many other schema libraries do it. However, it will be much harder to agree on a common structure specifically for Valibot. If we can't natively integrate the AST into our library code, it's probably not an option for us, as it would increase the bundle size a lot if we were forced to add it on top.

Going back to your initial suggestions with `~defaults`, I think the idea is great, but probably very form library specific. Personally, I would prefer that the user provide initial form values separately and not as part of the schema. I don't have a good example right now, but I remember that there are some edge cases where it's not a good idea to mix the two. Also, there are probably many questions about how to implement the `~defaults` property. How to handle dynamic defaults that use a generator function? How to handle unions? Should `~defaults` contain the recursively summed defaults of the whole schema (this would again be a problem in terms of bundle size for valibot) or just the default of a single schema as explicitly passed to `.default` in Zod and `optional` in Valibot, if defined.

Standard Schema is intentionally not able to cover every use case, so for more complicated and individual cases it will still be necessary to use adapters. If Standard Schema becomes too specific, we would block innovation in the long run, as new schema libraries that can't apply to this specification have no chance of being widely adopted if one day every third-party library only accepts Standard Schema.

**Comment by logaretm** [comment]
> and critically, is incompatible with unions.

Yep, I have no idea how would that work with a `~defaults` kind of behavior. While libraries like mine can afford to do best effort here, I think a standard spec will need to do better.

I didn't see a lot of unions/intersections with forms in the wild, so I'm unable to suggest a path forward with this point in particular.

> Realistically, we'd need to address https://github.com/standard-schema/standard-schema/issues/7

I just took a look at this and I agree. Having a standard AST will go along way in making things like this user-land or library-land in my case, as i can go through the AST and decide what I want to do for my users and can make certain compromises that a standard spec cannot.

> there are probably many questions about how to implement the ~defaults property...

Yep, the way we do them at the moment is different for each schema provider given the different tools available for each. For Zod we walk the AST and collect what we can. I totally agree from a standard standpoint this would be hard to implement. But given an AST, then you can leave those questions to the implementer.

ASTs also will solve an issue I have in mind with what I can path descriptions, which are simple flags/info about a given field. I will leave more details there on that issue.

Thanks for the feedback everyone, I will be annoying you with more Form specific stuff, this project is something I have been dreaming about for a while so I appreciate the time and work you are putting in here.
